### PR TITLE
feat: add provider allowlist support

### LINKED ISSUE #56: [Feature]: Provider Whitelist Support
Would it be possible to somehow enforce provider whitelisting?

For example, in one of the first sessions I was testing with, I had the following providers come up.

1. `hashicorp/aws` v6.42.0
2. `kuwas/github` v4.3.0
3. `integrations/github` v6.6.0

While I would have definitely wanted to work provider 1 + 3, 2 was absolutely not trusted by me.  By having some mechanism to limit `spacelift-intent` to a set of trusted providers it would make adoption of this tool a bit easier, not leaving the choice up purely do the user and LLM.

Having it as part of a prompt, just doesn't seem reliable as a restriction.

### PR DESCRIPTION
**Pull request by sephriot**
# Pull Request

## 📌 Reason

Today the MCP server accepts any provider name the LLM hands it. Trust decisions sit entirely in chat between the user and the model — a typosquatting / lookalike provider like `kuwas/github` can slip in alongside `hashicorp/aws` and `integrations/github`. Prompt-level guidance is unreliable. This PR adds a deployer-controlled allowlist so the agent cannot widen its own trust boundary at runtime.

## 📄 Summary

Adds `--provider-allowlist-file` (env `PROVIDER_ALLOWLIST_FILE`) that points at a YAML file or the sentinel `builtin:trusted`. When set, providers outside the list are rejected before any binary download, and stripped from `provider-search` results so the model never sees them. When unset, behavior is unchanged.

### Core Improvements

- New `allowlist/` package: YAML loader, matcher (exact + namespace wildcard, optional version constraint), embedded `builtin:trusted` set (`hashicorp/*` + `opentofu/*`).
- Hard gate in `provider.OpenTofuAdapter.LoadProvider` rejects denied providers before any registry interaction or GRPC plugin spawn.
- `registry.NewAllowlistedClient` decorator filters `SearchProviders` / `FindProvider` / `GetProviderVersions` / `GetProviderDownload` against the allowlist.
- MCP server `Instructions` are augmented with the configured allowlist summary so the model knows the constraint upfront.

### Technical Details

- Version constraint syntax follows Terraform's `required_providers` via `github.com/hashicorp/go-version` — users don't have to learn a second mini-language.
- Match semantics: most-specific-wins (exact name overrides namespace wildcard); within the same specificity tier, the provider is allowed if any entry permits its version (OR). Constraints do not inherit between rules.
- `Allowlist` is nil-safe; `Disabled()` and `nil` both mean "permit everything".
- `NewAllowlistedClient` returns the inner client unchanged when the allowlist is disabled — zero overhead in the default configuration.

## 🎯 Technical Details

**File format:**

```yaml
providers:
  - name: hashicorp/aws
    versions: ">= 5.0.0, < 7.0.0"   # optional
  - name: hashicorp/random
  - name: hashicorp/*
  - name: integrations/github
    versions: "~> 6.0"
```

**Failure modes:** flag set with a missing/malformed file or invalid constraint → fail-closed at startup with a clear error. No silent fallback.

**Out of scope (deferred):** cross-namespace wildcards (`*/aws`), shasum pinning, allowlist hot-reload, MCP tools to mutate the allowlist (deliberately omitted — the agent must not widen its own trust boundary).

## ✅ Testing

- [x] Unit tests pass — 43 new tests across `allowlist/`, `registry/`, `provider/`, `instructions/` covering match semantics edge cases (tier precedence, OR within tier, constraint non-inheritance), malformed input, builtin load, and registry decorator filtering.
- [x] `LoadProvider` gate verified with a recording fake registry: registry is never contacted on allowlist denial (name or version).
- [x] Manual testing: `--help` confirms the flag is wired; binary builds cleanly.
- [ ] Integration tests (`test/`) not re-run — they need credentials and are unaffected by this change (they construct the adapter with `nil` allowlist via `testhelper`).
- [x] No breaking changes — flag unset preserves today's behavior.

## 📋 Checklist

- [x] Code follows project conventions
- [x] Self-review completed
- [x] Documentation updated (README "Provider Allowlist" section + flags table)
- [x] Tests added/updated for new functionality
- [x] No sensitive information exposed
- [x] Related issues linked
- [ ] Error handling improved (n/a — new feature)
- [x] No breaking changes introduced

## 🚀 Impact

- **Default (no flag)**: zero behavioral change. Server logs a one-line warning at startup noting the allowlist is disabled.
- **With `builtin:trusted`**: quick opt-in to a curated default permitting `hashicorp/*` and `opentofu/*`.
- **With a custom file**: deployer-controlled trust boundary, immune to LLM-side prompt injection or mistakes.

The allowlist is intentionally loaded once at startup and has no MCP tools to mutate it — the trust boundary is owned by the deployer, not the agent.

## 🔗 Related Issues

Closes #56.

### CODE REVIEW

File: allowlist/allowlist.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+	enabled bool
+	entries []entry
+}
+
+type entry struct {
+	namespace string
+	// name is empty for a namespace wildcard ("hashicorp/*").
+	name string
+	// constraints is nil when no version restriction is configured.
+	constraints version.Constraints
```
Comment (Yantrio) [inline allowlist/allowlist.go:38]: love that you're using go-version version constraints for this, so it matches OpenTofu's logic here <3

File: registry/allowlist_client.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	if len(results) == 0 {
+		return nil, nil
+	}
+	best := &results[0]
+	for i := range results {
+		if results[i].Popularity > best.Popularity {
+			best = &results[i]
+		}
+	}
+	return best, nil
```
Comment (Yantrio) [inline registry/allowlist_client.go]: I wonder if we could use the inner.SearchProviders here instead?

Comment (sephriot) [reply registry/allowlist_client.go]: Done in f0600af. Now calls `inner.SearchProviders` directly and inlines the filter + popularity pick — clearer single pass instead of going through the wrapper's own `SearchProviders`.

**Review by Yantrio — APPROVED** [review]

**Review by michalrom089 — APPROVED** [review]

**Comment by sephriot** [comment]
Closing due to inactivity
